test/index.js

```diff
@@ -616,8 +616,9 @@ describe('def', function() {
 
     throws(function() { defaultTo(0, 'XXX'); },
            errorEq(TypeError,
-                   '‘defaultTo’ expected a value of type (Nullable Number) ' +
-                   'as its second argument; received "XXX"'));
+                   '‘defaultTo’ expected its first and second arguments to ' +
+                   'be of the same type; the first is of type Number' +
+                   ', however the second, "XXX", is of type String'));
```

Comment:
Alright, pushed some changes. If you pull down the branch and run `make test`, `supports "nullable" types` should fail and you should see our new error message. Most of the diff is creating a `signature` POJO that we can in theory use to print our signature and highlight certain parts (eg `sigShowTypeVarPositions`). I pretty much just tried to make as few changes as possible in this iteration to surrounding code. The implementation is mostly imperative.

The signature is made up of four parts (name, constraints, arguments, return type) which I generically have called "group"s. Each group can have 0 or more "item"s (ie a function that takes 2 arguments would have 2 items)

I haven't implemented constraints. I didn't really look into it, but I wasn't sure off the top of my head how we could derive something like `filter :: (Monad m, Monoid (m a)) => (a -> Boolean) -> m a -> m a`

Also, `_satisfactoryTypes` now takes 10 arguments, so we have that going for us :dancer: I think it might be nice to create a POJO in `def` that simply wraps the things the signature needs and pass that around, ie

```js
return curry(name,
             constraints,
             expArgTypes,
             expTypes[expTypes.length - 1],
             {},
             new Array(arity),
             range(0, arity),
             impl);
```
would become
```js
var _sig = {
  name: name,
  constraints: constraints,
  args: expArgTypes,
  ret: expTypes[expTypes.length - 1]
};
return curry(_sig, {}, new Array(arity), range(0, arity), impl);
```
I'm sure I had more to say about this all, but it's getting late :sleeping: Let me know what you think of this direction.